### ISSUE TITLE
High memory usage

### ISSUE THREAD

**Issue by g199209** [martinellimarco/libzstd-seek#7]
Test program:

```C++
int main() {
    int fd = open("xxxx.zst", O_RDONLY);
    auto* zstdCtx = ZSTDSeek_createFromFileDescriptor(fd);

    std::vector<std::byte> buf(1024 * 1024);
    while (true) {
        auto bytesRead = ZSTDSeek_read(buf.data(), buf.size(), zstdCtx);
        if (bytesRead == 0) break;
    }

    return 0;
}
```

The input sztd file is created by t2sz tool,  which is about 6.5G, and the raw uncompressed size is about 42G. 

When I run this program, the memory usage is increasing by time, and reach about 6.5G at the end, which is similar to the input file size.

Why this lib need so much memory? Is there any way to reduce the memory usage?

**Comment by martinellimarco** [comment]
This is expected behavior. it's how mmap() works, not a memory leak.

When libzstd-seek opens your file, it calls mmap() to map the entire 6.5GB compressed file into the process's virtual address space. No physical RAM is used at this point, it's just a mapping.

As your while(true) loop reads through the file, each new region triggers a page fault. The kernel loads that page from disk into RAM. By the time you've read all 42GB of uncompressed data, every page of the 6.5GB compressed file has been touched at least once, so they're all resident in memory.

Here's the key point: this is not allocated memory. These are file-backed, read-only pages, functionally identical to the OS page cache. The kernel can (and will) reclaim them instantly under memory pressure, with zero cost, because the data is already on disk.

You can verify this yourself: check VmRSS vs RssFile in /proc/<pid>/status. The ~6.5GB will show up entirely under RssFile (file-backed pages), not under anonymous memory. If you run the same program on a system with only 4GB of free RAM, it will still work fine, the kernel will simply evict older pages as it loads new ones.

The actual heap memory used by the library is just a few hundred KB (decompression context + jump table + temp buffer).

TL;DR: What you're seeing in top/htop is the OS page cache for the memory-mapped file, not real memory consumption. No changes needed.

**Comment by g199209** [comment]
> This is expected behavior. it's how mmap() works, not a memory leak.
> 
> When libzstd-seek opens your file, it calls mmap() to map the entire 6.5GB compressed file into the process's virtual address space. No physical RAM is used at this point, it's just a mapping.
> 
> As your while(true) loop reads through the file, each new region triggers a page fault. The kernel loads that page from disk into RAM. By the time you've read all 42GB of uncompressed data, every page of the 6.5GB compressed file has been touched at least once, so they're all resident in memory.
> 
> Here's the key point: this is not allocated memory. These are file-backed, read-only pages, functionally identical to the OS page cache. The kernel can (and will) reclaim them instantly under memory pressure, with zero cost, because the data is already on disk.
> 
> You can verify this yourself: check VmRSS vs RssFile in /proc//status. The ~6.5GB will show up entirely under RssFile (file-backed pages), not under anonymous memory. If you run the same program on a system with only 4GB of free RAM, it will still work fine, the kernel will simply evict older pages as it loads new ones.
> 
> The actual heap memory used by the library is just a few hundred KB (decompression context + jump table + temp buffer).
> 
> TL;DR: What you're seeing in top/htop is the OS page cache for the memory-mapped file, not real memory consumption. No changes needed.

Thanks for the explanation. I still think the “reclaimed instantly, so it will still work fine” statement is too strong.

In my real system, I run multiple reader processes concurrently, and I did hit OOM.  

Yes, these mmap pages are file-backed and *reclaimable in principle*, but reclaim is not a hard real-time guarantee. Under high concurrent memory pressure, there can be a reclaim lag/window where enough pages are not freed quickly enough, and the kernel enters OOM before reclaim catches up.

So the behavior I observed is: no obvious leak, but still OOM in practice when many mmap-based readers run together.

**Comment by martinellimarco** [comment]
Hi @g199209 , I haven't see your reply.

Can you give me more information about your use case? I'm interested in knowing how this library is used in practice and if you are encountering an OOM I may reconsider it and move away from mmap. No promises.

**Comment by g199209** [comment]
> Hi [@g199209](https://github.com/g199209) , I haven't see your reply.
> 
> Can you give me more information about your use case? I'm interested in knowing how this library is used in practice and if you are encountering an OOM I may reconsider it and move away from mmap. No promises.

Just as my test program show. You can construct a lot of large zstd files, and run a lot of (equals to CPU core count) programs concurrently.
